### ISSUE TITLE
perf(recurrent_gemma): decode runs at about a fifth of mlx-lm on M5 Max

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1689]
## Problem / Background

#1687 made the family work. Commit `f900a655` fixed three stacked defects: the conv1d NLC layout, RoPE applied before the head transpose, and a layer loop that iterated the repeating `block_types` pattern and so built 3 of 38 layers. Greedy output now matches mlx-lm token for token over 60 tokens of a chat prompt on `alpindale/recurrentgemma-9b`. This issue is purely about speed, not correctness.

## Current Behavior

Measured on `alpindale/recurrentgemma-9b` (17.96 GB, bf16, 38 layers, `lru_width` 4096, `conv1d_width` 4, 16 query heads over 1 KV head, `attention_window_size` 2048), at the repo's pp512/tg128 shape, `./scripts/bench_decode.sh <path> --max-tokens 128 --prompt-tokens 512` against `scripts/bench_mlxlm.py`:

| Phase | mlxcel | mlx-lm | Ratio |
|---|---|---|---|
| Decode | 5.74 tok/s | 29.48 tok/s | 0.19x |
| Prefill | 446.55 tok/s | 1645.73 tok/s | 0.27x |

## Evidence that this is not a structural porting mistake

Every checkpoint tensor shape matches what the port assumes, verified directly: `rg_lru.input_gate_weight` is `[16, 256, 256]`, `conv_1d.weight` is `[4096, 1, 4]`, `linear_x.weight` is `[4096, 4096]`, `q_proj.weight` is `[4096, 4096]`, `k_proj.weight` is `[256, 4096]`.

The RGLRU gate arithmetic, `apply_block_linear`, `rnn_scan` in both its sampling and linear modes, `Conv1d` padding and cache tail, `ResidualBlock` norm and residual ordering, the MLP, and the logits path with `logits_soft_cap` were each read against `mlx_lm/models/recurrent_gemma.py` and match.

Removing `repeat_kv` from `LocalAttentionBlock` measured within noise here: 5.68 to 5.73 tok/s at 64 prompt tokens, and 5.73 to 5.74 at 2048. This checkpoint has the largest expansion of any caller at `n_rep` 16, and on every other decoder that removal was worth 1.4x to 1.8x at 2048 tokens. That the same change is invisible here is itself evidence that something much larger dominates the step.

## Candidate for the prefill half

`rnn_scan`'s linear mode at `src/models/recurrent_gemma.rs:206-224` builds its output by pushing one array per timestep into a `Vec` and then folding them with repeated `concatenate` calls, which is quadratic in sequence length in graph nodes. That would explain the prefill number.

The decode half takes the `seq_len == 1` branch at `src/models/recurrent_gemma.rs:182` instead, so it needs its own explanation. Do not pin both halves on this one cause.

## Scope

**In scope:** profiling and optimizing the decode and prefill paths of `src/models/recurrent_gemma.rs`, and updating the `docs/supported-models.md` entry for the family.

**Out of scope:** correctness changes to the family, which #1687 already settled; changes to `repeat_kv` or shared attention helpers, since the measurement above shows they are not what dominates here.

## Acceptance Criteria

- [ ] A profile that attributes the decode cost to specific operations, rather than an assertion about the cause.
- [ ] Decode and prefill re-measured at the same pp512/tg128 shape and recorded, so the change can be compared against 5.74 and 446.55 tok/s.
- [ ] The `docs/supported-models.md` entry for the family is updated, since line 59 currently states the roughly 20% figure as a known gap.

## Verification

```bash
cargo build --release --features metal,accelerate
./scripts/bench_decode.sh <path-to-recurrentgemma-9b> --max-tokens 128 --prompt-tokens 512
.venv-mlxlm/bin/python scripts/bench_mlxlm.py --model <path-to-recurrentgemma-9b> --max-tokens 128 --prompt-tokens 512
cargo test --release recurrent_gemma
cargo clippy --all-targets -- -D warnings
```

A pass is a measured decode number recorded at pp512/tg128 next to the 5.74 tok/s baseline, plus greedy output still matching mlx-lm token for token over 60 tokens of a chat prompt.

## Technical Considerations

Environment: mlxcel 0.7.0-beta.1 on branch `bench/0.7.0-refresh`, MacBook Pro M5 Max 128GB, macOS 26.6.2, MLX pin 9a795735, built with `cargo build --release --features metal,accelerate`.

Related: #1687, commit `f900a655`.

**Comment by inureyes** [comment]
Correcting the prefill figure in the description. The 446.55 tok/s it cites was measured while a 26 GB checkpoint download was saturating the same disk, and it does not reproduce. Three clean runs at the same pp512/tg128 shape on an idle machine give 860.06, 862.77 and 866.40 tok/s, so prefill is about 864 and the ratio against mlx-lm's 1645.73 is roughly 53%, not 27%.

The decode figure stands. The same three runs give 5.65, 5.65 and 5.54 tok/s against 29.48, so decode is about 19% of the reference and that is the half worth chasing first.

This narrows the issue rather than widening it. The description offered the quadratic `concatenate` fold in `rnn_scan`'s linear mode as a prefill candidate and said the `seq_len == 1` branch needed its own explanation; with prefill at half the reference rather than a quarter, the decode branch is now clearly the dominant term. `docs/supported-models.md` has been corrected to the same numbers.

**Comment by inureyes** [comment]
Resolved by #1709. The cause was not in this model: `mlx_cxx_bridge.cpp`'s non-compiled activation helpers returned f32 for half-precision input, the promoted activation entered the residual stream, and from there every matmul in every later layer read its f16 weights promoted to f32. Griffin reaches that path twice per layer, at `recurrent_gemma.rs:342` and `:549`.

Measured on M5 Max at pp512/tg128 against `benchmarks/pylm_m5max_2026-09-06.csv`. Decode 5.54 to 29.23 tok/s, 5.28x, which is 99.2% of mlx-lm's 29.48. Prefill 866.40 to 1526.98, 93% of 1645.73. Effective bandwidth 108 to 557 GB/s, at the hardware's limit rather than a fifth of it. Greedy output is unchanged: the 60-token chat completion that already matched mlx-lm token for token still does, LaTeX block included.

Three hypotheses this issue raised were tested and rejected, which is worth recording so nobody re-runs them. The quadratic `concatenate` fold in `rnn_scan`'s linear mode was named as a prefill candidate; prefill was in fact at 53% of the reference, not 27%, because the figure in the description was measured while a 26 GB download saturated the same disk. Replacing the per-layer cache `copy` with a `take` removed 52 array copies per token and measured 5.54 to 5.65, noise. Folding `softplus(recurrent_param)` and the `-8.0` factor to load time removed about 208 ops per token and measured 2.9% before #1709 and nothing after it, because that apparent gain was dispatch weight in the f32-promoted regime rather than a real saving.

Nothing model-specific remains open here, so closing. #1710 covers the separate f32 attention guard that #1709's work surfaced in `phi` and `stablelm`.
